python/python/lance/torch/kmeans.py

```diff
@@ -0,0 +1,189 @@
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import logging
+from typing import List, Optional, Union
+
+import numpy as np
+import pyarrow as pa
+import torch
+from tqdm import tqdm
+
+from . import preferred_device
+from .distance import cosine_distance, dot_distance, l2_distance
+
+__all__ = ["KMeans"]
+
+
+@torch.jit.script
+def _random_init(
+    data: torch.Tensor, n: int, seed: Optional[int] = None
+) -> torch.Tensor:
+    if seed is not None:
+        torch.random.manual_seed(seed)
+    sample_idx = torch.randint(0, data.shape[0], (n,))
+    samples = data[sample_idx]
+    return samples
+
+
+class KMeans:
+    """K-Means trains over vectors and divide into K clusters.
+
+    This implement is built on PyTorch, supporting CPU, GPU and Apple Silicon GPU.
+
+    Parameters
+    ----------
+    k: int
+        The number of clusters
+    metric : str
+        Metric type, support "l2", "cosine" or "dot"
+    init: str
+        Initialization method. Only support "random" now.
+    max_iters: int
+        Max number of iterations to train the kmean model.
+    tolerance: float
+        Relative tolerance with regards to Frobenius norm of the difference in
+        the cluster centers of two consecutive iterations to declare convergence.
+    seed: int, optional
+        Random seed
+    device: str, optional
+        The device to run the PyTorch algorithms. Default we will pick
+        the most performant device on the host. See `lance.torch.preferred_device()`
+    """
+
+    def __init__(
+        self,
+        k: int,
+        *,
+        metric: str = "l2",
+        init: str = "random",
+        max_iters: int = 50,
+        tolerance: float = 1e-4,
+        seed: Optional[int] = None,
+        device: Optional[str] = None,
+    ):
+        self.k = k
+        self.max_iters = max_iters
+
+        metric = metric.lower()
+        self.metric = metric
+        if metric in ["l2", "euclidean"]:
+            self.dist_func = l2_distance
+        elif metric == "cosine":
+            self.dist_func = cosine_distance
+        elif metric == "dot":
+            self.dist_func = dot_distance
+        else:
+            raise ValueError(
+                f"Only l2/cosine/dot is supported as metric type, got: {metric}"
+            )
+
+        self.centroids: Optional[torch.Tensor] = None
+        self.init = init
+        self.device = preferred_device(device)
+        self.tolerance = tolerance
+        self.seed = seed
+
+    def __repr__(self):
+        return f"KMeans(k={self.k}, metric={self.metric}, device={self.device})"
+
+    def _to_tensor(
+        self, data: Union[pa.FixedSizeListArray, np.ndarray, torch.Tensor]
+    ) -> torch.Tensor:
+        if isinstance(data, pa.FixedSizeListArray):
+            data = torch.from_numpy(np.stack(data.to_numpy(zero_copy_only=False)))
```

Comment:
At some point we'll want to switch to passing buffers, but this needs to be tested.
```python
import pyarrow as pa
import torch

la = pa.FixedSizeListArray.from_arrays([[0,1],[3,4],[5, 6]], 3)
torch.frombuffer(la.buffers()[4], dtype=torch.int64).reshape(3,2)
```

Reply:
sg